src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,149 @@
+            this.prev = prev;
+            this.value = value;
+            this.next = next;
+        }
+    }
+
     @Override
     public void add(T value) {
+        if (isEmpty()) {
+            head = new Node<>(null, value, null);
+            tail = head;
+        } else {
+            Node<T> nextNode = new Node(tail, value, null);
+            tail.next = nextNode;
+            tail = nextNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size) {
+            add(value);
+            return;
+        } else if (index == 0) {
+            Node<T> nextNode = new Node<>(null, value, head);
+            head.prev = nextNode;
+            head = nextNode;
+        } else {
+            checkIndex(index);
+            Node<T> targetNode = getNode(index);
+            Node<T> newNode = new Node<>(targetNode.prev, value, targetNode);
+            targetNode.prev.next = newNode;
+            targetNode.prev = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return getNode(index).value;
+    }
+
+    void checkIndex(int index) {
+        if (index < 0 | index >= size) {
+            throw new IndexOutOfBoundsException("Index isn't valid.");
+        }
+    }
+
+    Node<T> getNode(int index) {
+        Node<T> node;
+        if (index < size / 2) {
+            node = head;
+            for (int i = 0; i < index; i++) {
+                node = node.next;
+            }
+        } else {
+            node = tail;
+            for (int i = size - 1; i > index; i--) {
+                node = node.prev;
+            }
+        }
+        return node;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> targetNode = getNode(index);
+        T returnValue = targetNode.value;
+        targetNode.value = value;
+        return returnValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> targetNode = getNode(index);
+        return unlink(targetNode);
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> node = head;
+        if (object == null) {
+            while (node != null) {
+                if (node.value == null) {
+                    unlink(node);
+                    return true;
+                }
+                node = node.next;
+            }
+        } else {
+            while (node != null) {
+                if (node.value.equals(object)) {
+                    unlink(node);
+                    return true;
+                }
+                node = node.next;
+            }
+        }
```

Comment:
we don't need this if-else construct and we can leave only one while loop if we change our if statement inside this while loop like this:

if (object == node.value || (object != null && object.equals(node.value)))